source/common/stats/thread_local_store.h

```diff
@@ -16,6 +16,77 @@
 namespace Envoy {
 namespace Stats {
 
+/**
+ * Log Linear Histogram implementation per thread.
+ */
+class ThreadLocalHistogramImpl : public Histogram, public MetricImpl {
+public:
+  ThreadLocalHistogramImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                           std::vector<Tag>&& tags);
+
+  virtual ~ThreadLocalHistogramImpl();
+  // Stats::Histogram
+  void recordValue(uint64_t value) override;
+
+  // TODO(ramaraochavali): split the Histogram interface in to two - parent and tls.
+  void merge() override { NOT_IMPLEMENTED; }
+  bool used() const override { return flags_ & Flags::Used; }
+  const HistogramStatistics& intervalStatistics() const override { return interval_statistics_; }
+  const HistogramStatistics& cumulativeStatistics() const override {
+    return cumulative_statistics_;
+  }
+
+  void beginMerge() { current_active_ = 1 - current_active_; }
+  void merge(histogram_t* target);
+
+  Store& parent_;
+
+private:
+  uint64_t current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+  HistogramStatisticsImpl interval_statistics_;
+  HistogramStatisticsImpl cumulative_statistics_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Log Linear Histogram implementation that is stored in the main thread.
+ */
+class HistogramParentImpl : public Histogram, public MetricImpl {
+public:
+  HistogramParentImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                      std::vector<Tag>&& tags);
+
+  virtual ~HistogramParentImpl();
+
+  // TODO(ramaraochavali): split the Histogram interface in to two - parent and tls.
+  void recordValue(uint64_t) override { NOT_IMPLEMENTED; }
+  bool used() const override;
+  void merge() override;
+  const HistogramStatistics& intervalStatistics() const override { return interval_statistics_; }
+  const HistogramStatistics& cumulativeStatistics() const override {
+    return cumulative_statistics_;
+  }
+
+  void addTlsHistogram(TlsHistogramSharedPtr hist_ptr);
+
+  Store& parent_;
+  std::list<TlsHistogramSharedPtr> tls_histograms_;
+
+private:
+  bool usedWorker() const;
+
+  histogram_t* interval_histogram_;
+  histogram_t* cumulative_histogram_;
+  HistogramStatisticsImpl interval_statistics_;
+  HistogramStatisticsImpl cumulative_statistics_;
+  mutable std::mutex merge_lock_;
```

Comment:
Could you point me at docs of the threading model?  I see we're really good about grabbing merge_lock_ before doing any work on the histograms but it concerns me that callers can just grab raw pointers outside the locks

Reply:
Threading model was I think based on @mattklein123 issue comment: https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282

It's probably worth pointing to that in the code itself.

Reply:
ok. I have actually added them at implementation. I moved the doc pointer to interface now.